cpp/tests/io/parquet_chunked_reader_test.cpp

```diff
@@ -0,0 +1,888 @@
+  // table rows.
+  if (nullable) {
+    // Generate deterministic bitmask instead of random bitmask for easy computation of data size.
+    auto const valid_iter = cudf::detail::make_counting_transform_iterator(
+      0, [](cudf::size_type i) { return i % 4 == 3 ? 0 : 1; });
+
+    cudf::size_type offset{0};
+    for (auto& col : input_columns) {
+      auto const col_typeid = col->type().id();
+      col->set_null_mask(
+        cudf::test::detail::make_null_mask(valid_iter + offset, valid_iter + col->size() + offset));
+
+      if (col_typeid == cudf::type_id::STRUCT) {
+        auto const null_mask  = col->view().null_mask();
+        auto const null_count = col->null_count();
+
+        for (cudf::size_type idx = 0; idx < col->num_children(); ++idx) {
+          cudf::structs::detail::superimpose_parent_nulls(null_mask,
+                                                          null_count,
+                                                          col->child(idx),
+                                                          cudf::get_default_stream(),
+                                                          rmm::mr::get_current_device_resource());
+        }
+      }
+
+      // Can't use `cudf::detail::purge_nonempty_nulls` since it requires to be compiled with CUDA.
+      if (col_typeid == cudf::type_id::LIST) {
+        col = cudf::purge_nonempty_nulls(cudf::lists_column_view{col->view()});
+      } else if (col_typeid == cudf::type_id::STRUCT) {
+        col = cudf::purge_nonempty_nulls(cudf::structs_column_view{col->view()});
+      } else if (col_typeid == cudf::type_id::STRING) {
+        col = cudf::purge_nonempty_nulls(cudf::strings_column_view{col->view()});
+      }
+    }
+  }
+
+  auto input_table = std::make_unique<cudf::table>(std::move(input_columns));
+  auto filepath =
+    temp_env->get_temp_filepath(nullable ? filename + "_nullable.parquet" : filename + ".parquet");
+
+  auto const write_opts =
+    cudf::io::parquet_writer_options::builder(cudf::io::sink_info{filepath}, *input_table)
+      .max_page_size_bytes(max_page_size_bytes)
+      .max_page_size_rows(max_page_size_rows)
+      .build();
+  cudf::io::write_parquet(write_opts);
+
+  return std::pair{std::move(input_table), std::move(filepath)};
+}
+
+auto chunked_read(std::string const& filepath, std::size_t byte_limit)
+{
+  auto const read_opts =
+    cudf::io::parquet_reader_options::builder(cudf::io::source_info{filepath}).build();
+  auto reader = cudf::io::chunked_parquet_reader(byte_limit, read_opts);
+
+  auto num_chunks = 0;
+  auto result     = std::make_unique<cudf::table>();
+
+  do {
+    auto chunk = reader.read_chunk();
+    if (num_chunks == 0) {
+      result = std::move(chunk.tbl);
+    } else {
+      CUDF_EXPECTS(chunk.tbl->num_rows() != 0, "Number of rows in the new chunk is zero.");
+      result = cudf::concatenate(std::vector<cudf::table_view>{result->view(), chunk.tbl->view()});
```

Comment:
Could we concatenate once, after the loop? In the current code we copy chunks ~n^2 times.